### PR TITLE
docs: Add documentation for six pylint errors

### PR DESCRIPTION
**Pull request by ollie-iterators**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |
| ✓   | :scroll: Docs          |

## Description

This adds documentation for the following errors in pylint from pylint-errors:

- bad-mcs-classmethod-argument
- bad-mcs-method-argument
- inconsistent-mro
- docstring-first-line-empty
- differing-param-doc
- differing-type-doc

Refs [#5953](https://github.com/PyCQA/pylint/issues/5953)

Closes #XXXX

### CODE REVIEW

File: doc/data/messages/b/bad-mcs-method-argument/details.rst
Comment (Pierre-Sassoulas) [inline doc/data/messages/b/bad-mcs-method-argument/details.rst:7]: There's a description in the message directly in the checker, you should probably modify that and regenerate the doc :)

File: doc/whatsnew/fragments/0203.docs
Comment (Pierre-Sassoulas) [inline doc/whatsnew/fragments/0203.docs:1]: No need to document this with fragments, I added the skip news label

File: doc/data/messages/d/differing-param-doc/bad.py
Comment (Pierre-Sassoulas) [inline doc/data/messages/d/differing-param-doc/bad.py:1]: Could you use real world example without foo / bar, please ? It's more lively and understandable for beginner with actual examples.

File: doc/data/messages/b/bad-mcs-method-argument/details.rst
Comment (ollie-iterators) [reply doc/data/messages/b/bad-mcs-method-argument/details.rst]: Changed documentation to what I found in "doc/user_guide/checkers/features.rst"

File: doc/whatsnew/fragments/0203.docs
Comment (ollie-iterators) [reply doc/whatsnew/fragments/0203.docs]: Removed the fragments

File: doc/data/messages/d/differing-param-doc/bad.py
Comment (ollie-iterators) [reply doc/data/messages/d/differing-param-doc/bad.py]: Changed to a real word example

File: doc/data/messages/d/differing-param-doc/bad.py
Comment (Pierre-Sassoulas) [inline doc/data/messages/d/differing-param-doc/bad.py:1]:
```suggestion
def leap_year(year):
```
Let's respect pep8 / python norms as much as we can :)

File: doc/data/messages/d/differing-param-doc/good.py
Comment (Pierre-Sassoulas) [inline doc/data/messages/d/differing-param-doc/good.py:1]:
```suggestion
def leap_year(year):
```

File: doc/data/messages/b/bad-mcs-method-argument/details.rst
Comment (Pierre-Sassoulas) [inline doc/data/messages/b/bad-mcs-method-argument/details.rst:4]: Check the generated result, we're already using what's in the checker's message's description automatically. If there's nothing to add we can remove the details.rst altogether

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: doc/data/messages/d/differing-param-doc/bad.py
Comment (ollie-iterators) [reply doc/data/messages/d/differing-param-doc/bad.py]: Fixed

File: doc/data/messages/d/differing-param-doc/good.py
Comment (ollie-iterators) [reply doc/data/messages/d/differing-param-doc/good.py]: Fixed

File: doc/data/messages/b/bad-mcs-method-argument/details.rst
Comment (ollie-iterators) [reply doc/data/messages/b/bad-mcs-method-argument/details.rst]: The generated result from pylint was "Metaclass class method __new__ should have 'mcs' as first argumentPylint(C0204:bad-mcs-classmethod-argument)"

**Comment by Pierre-Sassoulas** [comment]
You can generate the doc locally with ``tox -e docs``

**Comment by ollie-iterators** [comment]
> You can generate the doc locally with `tox -e docs`

I just did that

File: doc/data/messages/b/boolean-datetime/bad.py
Comment (Pierre-Sassoulas) [inline doc/data/messages/b/boolean-datetime/bad.py]: We'd need an example of bad code for this message

File: doc/data/messages/b/bad-mcs-method-argument/bad.py
Comment (Pierre-Sassoulas) [inline doc/data/messages/b/bad-mcs-method-argument/bad.py:3]: There need to be a message raised by the bad exampele somewhere, you can launch the check with ``pytest doc -k bad-mcs``

File: doc/data/messages/c/cyclic-import/bad.py
Comment (Pierre-Sassoulas) [inline doc/data/messages/c/cyclic-import/bad.py:1]: This is not valid python, it can't work. I think cyclic-import is not the easiest message to document as it implicate multiple files.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: doc/data/messages/c/cyclic-import/bad.py
Comment (ollie-iterators) [reply doc/data/messages/c/cyclic-import/bad.py:1]: I got the documentation from pylint-errors, which is where I was supposed to get it from for: https://github.com/PyCQA/pylint/issues/5953

File: doc/data/messages/b/boolean-datetime/bad.py
Comment (ollie-iterators) [reply doc/data/messages/b/boolean-datetime/bad.py]: The error doesn't happen unless you use Python version 3.5

Comment (Pierre-Sassoulas) [reply doc/data/messages/b/boolean-datetime/bad.py]: Use py-version=3.5 like us f-string-used- ... in version that does not support it.

File: doc/data/messages/c/cyclic-import/bad.py
Comment (Pierre-Sassoulas) [reply doc/data/messages/c/cyclic-import/bad.py:1]: We added some more checks and normalization around it. Most of the message are already implemented so the remaining one can be a little hard to do (like cyclic-import or duplicate code). It's ok if you don't want /can't handle the hard one :)

File: doc/data/messages/b/bad-mcs-method-argument/bad.py
Comment (ollie-iterators) [reply doc/data/messages/b/bad-mcs-method-argument/bad.py:3]: I found a message for the code

Comment (ollie-iterators) [reply doc/data/messages/b/bad-mcs-method-argument/bad.py:3]: I made an issue happen

File: doc/data/messages/b/boolean-datetime/bad.py
Comment (ollie-iterators) [reply doc/data/messages/b/boolean-datetime/bad.py]: Ok, I stopped changing boolean-datetime

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
The documentation check fail seems genuine, could you look into it please ?

**Comment by Pierre-Sassoulas** [comment]
Alternatively there's some of the check that can be merged already, if you create a new MR with only those it could prevent duplicate like in #8184

**Comment by ollie-iterators** [comment]
> Alternatively there's some of the check that can be merged already, if you create a new MR with only those it could prevent duplicate like in #8184

I removed the documentation for cyclic-import. I assume that is the only one with the issue after I removed boolean-datetime as well.

**Comment by ollie-iterators** [comment]
What is wrong with the base dependencies?

**Comment by Pierre-Sassoulas** [comment]
Relaunched, that's a temporary github issue

**Comment by ollie-iterators** [comment]
> Relaunched, that's a temporary github issue

So should I be able to commit the changes?

**Comment by Pierre-Sassoulas** [comment]
It seems there's a genuine issue in the current version, you can use ``pytest doc`` to verify locally.

**Comment by ollie-iterators** [comment]
When I ran that, I got errors for things that I did not work with, like differing-param-doc

**Comment by ollie-iterators** [comment]
This is what showed up:
FAILED doc/test_messages_documentation.py::test_code_examples[inconsistent-mro-bad] - AssertionError: assert Counter() == Counter({(9, ...ent-mro'): 1})
FAILED doc/test_messages_documentation.py::test_code_examples[too-many-public-methods-bad] - AssertionError: There should be at least one warning raised for...
FAILED doc/test_messages_documentation.py::test_code_examples[missing-format-string-key-bad] - AssertionError: There should be at least one warning raised f...
FAILED doc/test_messages_documentation.py::test_code_examples[docstring-first-line-empty-bad] - AssertionError: There should be at least one warning raised ...
FAILED doc/test_messages_documentation.py::test_code_examples[differing-param-doc-bad] - AssertionError: There should be at least one warning raised for 'ba...
FAILED doc/test_messages_documentation.py::test_code_examples[differing-type-doc-bad] - AssertionError: There should be at least one warning raised for 'bad...
FAILED doc/test_messages_documentation.py::test_code_examples[wrong-spelling-in-docstring-good] - SystemExit: 32
FAILED doc/test_messages_documentation.py::test_code_examples[wrong-spelling-in-docstring-bad] - SystemExit: 32
FAILED doc/test_messages_documentation.py::test_code_examples[wrong-spelling-in-comment-good] - SystemExit: 32
FAILED doc/test_messages_documentation.py::test_code_examples[wrong-spelling-in-comment-bad] - SystemExit: 32

**Comment by Pierre-Sassoulas** [comment]
The real one are:
 FAILED doc/test_messages_documentation.py::test_code_examples[differing-param-doc-bad]
FAILED doc/test_messages_documentation.py::test_code_examples[docstring-first-line-empty-bad]
FAILED doc/test_messages_documentation.py::test_code_examples[differing-type-doc-bad]
FAILED doc/test_messages_documentation.py::test_code_examples[missing-format-string-key-bad]
FAILED doc/test_messages_documentation.py::test_code_examples[too-many-public-methods-bad]
FAILED doc/test_messages_documentation.py::test_code_examples[inconsistent-mro-bad]

For the spelling you'd need to install pyenchant for it to work, you can do that or ignore it.

**Comment by ollie-iterators** [comment]
What is wrong with the errors? I have the wrong variable name in the differing-param-doc docstring.

File: doc/data/messages/d/differing-param-doc/pylintrc
Comment (Pierre-Sassoulas) [inline doc/data/messages/d/differing-param-doc/pylintrc:2]:
```suggestion
differing-param-doc=yes
```

File: doc/data/messages/d/docstring-first-line-empty/pylintrc
Comment (Pierre-Sassoulas) [inline doc/data/messages/d/docstring-first-line-empty/pylintrc:2]:
```suggestion
docstring-first-line-empty=yes
```

File: doc/data/messages/m/missing-format-string-key/pylintrc
Comment (Pierre-Sassoulas) [inline doc/data/messages/m/missing-format-string-key/pylintrc:2]:
```suggestion
missing-format-string-key=yes
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
pylintrc format is ini, not sure ``:`` works

File: doc/data/messages/d/docstring-first-line-empty/pylintrc
Comment (ollie-iterators) [reply doc/data/messages/d/docstring-first-line-empty/pylintrc]: So why is this still failing? It literally has the first line empty in the docstring

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
You also need to load the extension raising the message, for example ``load-plugins = pylint.extensions.docparams``

**Comment by Pierre-Sassoulas** [comment]
> And where do I find that information on that? I didn't see any information on that on the moving errors from pylint-errors to pylint issue.

It's in the page you're generating https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/missing-any-param-doc.html (You should probably generate locally with tox -e docs and check the result in the generated html)

**Comment by ollie-iterators** [comment]
Also, where is the information about how to specify when an error shows up in pylint in the documentation?

**Comment by ollie-iterators** [comment]
> > And where do I find that information on that? I didn't see any information on that on the moving errors from pylint-errors to pylint issue.
> 
> It's in the page you're generating https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/missing-any-param-doc.html (You should probably generate locally with tox -e docs and check the result in the generated html)

I have added that information to some of the pylint errors that I have added documentation for, and nothing is happening.

**Review by Pierre-Sassoulas — COMMENTED** [review]
Please do not pile new messages on top of the other. I'd rather have 12 merge requests and merge them one by one rather than this.

**Comment by ollie-iterators** [comment]
> Please do not pile new messages on top of the other. I'd rather have 12 merge requests and merge them one by one rather than this.

I'll remove the extra onces once it is explained how to fix the following error documentations:

- bad-mcs-classmethod-argument
- bad-mcs-method-argument
- differing-param-doc
- differing-type-doc
- docstring-first-line-empty
- inconsistent-mro

**Comment by ollie-iterators** [comment]
> Please do not pile new messages on top of the other. I'd rather have 12 merge requests and merge them one by one rather than this.

Maybe you should explain better how to make issues with the docstrings work with pylint.

File: doc/data/messages/b/bad-mcs-classmethod-argument/details.rst
Comment (Pierre-Sassoulas) [inline doc/data/messages/b/bad-mcs-classmethod-argument/details.rst:8]: I don't understand this comment. If something seems odd to you, then open an issue so we can discuss it. It makes no sense to add your personal dissent to the documentation.

File: doc/data/messages/b/bad-mcs-classmethod-argument/details.rst
Comment (Pierre-Sassoulas) [inline doc/data/messages/b/bad-mcs-classmethod-argument/details.rst:4]: This should be in the code description of bad-mcs-classmethod in the checker that will automatically end up in the documentation.

File: doc/data/messages/b/bad-mcs-method-argument/details.rst
Comment (Pierre-Sassoulas) [inline doc/data/messages/b/bad-mcs-method-argument/details.rst:4]: Same remark than for  bad-mcs-classmethod

File: doc/data/messages/b/broken-collections-callable/pylintrc
Comment (Pierre-Sassoulas) [inline doc/data/messages/b/broken-collections-callable/pylintrc:1]: I don't think this should be modified ?

File: doc/data/messages/i/inconsistent-mro/details.rst
Comment (Pierre-Sassoulas) [inline doc/data/messages/i/inconsistent-mro/details.rst:1]: I suppose the message description already say this

File: doc/user_guide/configuration/all-options.rst
Comment (Pierre-Sassoulas) [inline doc/user_guide/configuration/all-options.rst:174]:
```suggestion
**Default:**  ``(3, 10)``
```

File: doc/user_guide/configuration/all-options.rst
Comment (Pierre-Sassoulas) [inline doc/user_guide/configuration/all-options.rst:267]:
```suggestion
   py-version = [3, 10]
```

File: doc/user_guide/configuration/all-options.rst
Comment (Pierre-Sassoulas) [inline doc/user_guide/configuration/all-options.rst:1270]:
```suggestion
*Spelling dictionary name. Available dictionaries: en (aspell), en_AU (aspell), en_CA (aspell), en_GB (aspell), en_US (aspell).*
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you for lowering the number of message :)

File: doc/data/messages/b/broken-collections-callable/pylintrc
Comment (Pierre-Sassoulas) [reply doc/data/messages/b/broken-collections-callable/pylintrc:1]:
```suggestion
[main]
```

**Comment by ollie-iterators** [comment]
What is the issue with the docstring errors?

File: doc/data/messages/d/differing-param-doc/pylintrc
Comment (DanielNoord) [inline doc/data/messages/d/differing-param-doc/pylintrc:2]: This is the incorrect format. Please look at how other examples do this.

In this case this should work:
```suggestion
load-extension = pylint.extensions.docparams
```

**Comment by Pierre-Sassoulas** [comment]
Please stop changing the title, please just focus on the 3 messages that were passing the tests and take the review remarks into account until we merge, then open another merge request for the remaining messages.

**Comment by ollie-iterators** [comment]
Fine, if you are going to not provide any help.

File: doc/data/messages/d/differing-param-doc/pylintrc
Comment (ollie-iterators) [reply doc/data/messages/d/differing-param-doc/pylintrc]: Still showing errors

**Comment by ollie-iterators** [comment]
> Please stop changing the title, please just focus on the 3 messages that were passing the tests and take the review remarks into account until we merge, then open another merge request for the remaining messages.

No

**Comment by DanielNoord** [comment]
> > Please stop changing the title, please just focus on the 3 messages that were passing the tests and take the review remarks into account until we merge, then open another merge request for the remaining messages.
> 
> If you are going to be no help.

Please refrain from such harsh comments. We are all volunteers. We appreciate every minute somebody spends on making `pylint` or its documentation better, but spamming us with commits and false claims that things are broken only costs us time without bringing the PR any further.

I have fixed all 6 messages for now, but I think a review is still necessary as some are not really in the explicit style of some of the other examples. Also please stop merging all of your commits into the branch. It makes no sense to have this branch have 140 commits and just makes reviewing a pain.

**Comment by ollie-iterators** [comment]
> > > Please stop changing the title, please just focus on the 3 messages that were passing the tests and take the review remarks into account until we merge, then open another merge request for the remaining messages.
> > 
> > 
> > If you are going to be no help.
> 
> Please refrain from such harsh comments. We are all volunteers. We appreciate every minute somebody spends on making `pylint` or its documentation better, but spamming us with commits and false claims that things are broken only costs us time without bringing the PR any further.
> 
> I have fixed all 6 messages for now, but I think a review is still necessary as some are not really in the explicit style of some of the other examples. Also please stop merging all of your commits into the branch. It makes no sense to have this branch have 140 commits and just makes reviewing a pain.

Ok, fine

**Comment by ollie-iterators** [comment]
Why is missing-parentheses-for-call-in-test suddenly calling an error?

**Comment by DanielNoord** [comment]
@ollie-iterators I have no idea what you are referring to and I'm going to close this PR. You blatantly ignored my previous comment about not keeping all commits around and this PR is getting out of hand.

I created https://github.com/PyCQA/pylint/pull/8202 to not let this work go to waste and all commits credit you as they should. You are welcome to create another PR with more examples if you wish but please make sure tests pass locally before opening the PR so not to consume to much of the limited time of maintainers.

**Pull request closed without merging**
